payments/voucher_manager.go

```diff
@@ -0,0 +1,136 @@
+package payments
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/statechannels/go-nitro/types"
+)
+
+// paymentStatus stores the status of payments for a given payment channel.
+type paymentStatus struct {
+	channelSender   common.Address
+	channelReceiver common.Address
+	startingBalance *big.Int
+	largestVoucher  Voucher
+	currentBalance  Balance
+}
+
+// VoucherManager receives and generates vouchers. It is responsible for storing vouchers.
+type VoucherManager struct {
+	channels map[types.Destination]*paymentStatus
+	me       common.Address
+}
+
+// NewVoucherManager creates a new voucher manager
+func NewVoucherManager(me types.Address) *VoucherManager {
+	channels := make(map[types.Destination]*paymentStatus)
+	return &VoucherManager{channels, me}
+}
+
+// Register registers a channel for use, given the sender and starting balance of the channel
+func (vm VoucherManager) Register(channelId types.Destination, sender common.Address, receiver common.Address, startingBalance *big.Int) error {
+
+	balance := Balance{big.NewInt(0).Set(startingBalance), &big.Int{}}
+	voucher := Voucher{ChannelId: channelId, Amount: big.NewInt(0)}
+	data := &paymentStatus{sender, receiver, big.NewInt(0).Set(startingBalance), voucher, balance}
+	if _, ok := vm.channels[channelId]; ok {
+		return fmt.Errorf("channel already registered")
+	}
+
+	vm.channels[channelId] = data
+
+	return nil
+}
+
+// Remove deletes the channel's status
+func (vm *VoucherManager) Remove(channelId types.Destination) {
+	delete(vm.channels, channelId)
+}
+
+// Pay will deduct amount from balance and add it to paid, returning a signed voucher for the
+// total amount paid.
+func (vm *VoucherManager) Pay(channelId types.Destination, amount *big.Int, pk []byte) (Voucher, error) {
+	pStatus, ok := vm.channels[channelId]
+
+	voucher := Voucher{Amount: &big.Int{}}
+	if !ok {
+		return voucher, fmt.Errorf("channel not found")
+	}
+
+	if types.Gt(amount, pStatus.currentBalance.Remaining) {
+		return voucher, fmt.Errorf("unable to pay amount: insufficient funds")
+	}
+
+	pStatus.currentBalance.Remaining.Sub(pStatus.currentBalance.Remaining, amount)
+	pStatus.currentBalance.Paid.Add(pStatus.currentBalance.Paid, amount)
```

Comment:
Do we need to update `pStatus.voucher` here?

Reply:
Good catch!

Reply:
Fixed in https://github.com/statechannels/go-nitro/pull/844/commits/c76786eec9851a27b5642c92e79468a072a3fce3